### ISSUE TITLE
Loading acquisition dates blocks browser interaction

### ISSUE THREAD

**Issue by jisantuc** [radiantearth/stac-browser#69]
Loading acquisition dates for items blocks browser interaction.

Replication:
- load https://stacindex.org/catalogs/world-bank-light-every-night#/2Y8qvpmgGXMkKYTC6V1qskx8HMfNqp5KvprH?t=items
- try to use the refresh button while acquisition dates are being populated
- note that your refresh isn't honored until after the acquisition date loads are all complete

I replicated on Windows 10 + FF and + Chrome.

**Comment by jisantuc** [comment]
Copied from stac-utils/stac-index#14

**Comment by m-mohr** [comment]
I did some experiments today with Web Workers, but have issues to actually load axios and stac-migrate in the workers. 
It also seems most of the delay is not coming directly from the HTTP requests, but instead from Vue's reactivity and updating mechanism, which we won't be able to improve a lot. Have to look into it again later once STAC Browser v3 is more stable...

**Comment by m-mohr** [comment]
I had an idea this morning how to fix this and implemented it without web workers. Right now it seems to work fine, let's see whether it helps enough. 

Does it work for you, too @jisantuc ?
You can try it here: https://radiantearth.github.io/stac-browser (e.g. the Google Earth Engine has a lot of catalogs to load)

**Comment by jisantuc** [comment]
What am I supposed to do to test? I tried a few links from the external links section and they're all 404-ing. It looks like you might have lost a `:` in an interpolated string somewhere -- I'm getting an error that says `Requested URL: https//data.geo.admin.ch/api/stac/v0.9/` (not `https://`)

**Comment by m-mohr** [comment]
@jisantuc Oh, sorry, a regression. Should be fixed now (refresh cache if you still get the wrong URL), could you try again, please? 
When accessing the World Band Light Every Night Dataset I now keep a somewhat responsive browser window which I can still interact with. Does it work for you, too?

**Comment by jisantuc** [comment]
When you say World Band Light Every Night, are you referring to the original link? Also, what do you mean by "somewhat responsive"? I'm still unable to do things like highlight text or hit the refresh button at that link until all of the acquisition dates have been loaded.

**Comment by m-mohr** [comment]
No, the original link is still the old STAC Browser.

The updated version is here: https://radiantearth.github.io/stac-browser - paste the URL into the URL field and load the page.
It's still pretty unresponsive, but I can use the stop/reload buttons and click links (with a small delay), which was not possible before. Selecting text is indeed not really possible.

But we can only improve that by using Web Workers, which have their own issues and I don't have enough experience yet to solve them. World Band is in the end a poorly modeled catalog (for browsing purposes). STAC Browser has to transfer, store and process dozens of megabytes of data because there are thousands of entries in each catalog. That will never be responsive, except I disable preloading/loading details like acquisition dates completely.

**Comment by jisantuc** [comment]
Loading a moment ago felt much better. Mouse interactions like scroll working normally during load is a significant ergonomic improvement. I'm a little curious about one of the things you said though:

> We can only improve that (not solve it entirely) by using Web Workers

Is your claim that it's normal for pending http requests to block interaction outside of applications that push a bunch of work to web workers?

**Comment by m-mohr** [comment]
Thanks for checking, good to hear you've also noticed an improvement.

No, that's not my claim.
Some context: STAC Browser doesn't only request data, it requests data for (on my screen) 12 to 15 items/catalogs simultaneously. Some are very large and this is an issue as STAC Browser also needs to process those. All items/catalogs need to be migrated (i.e. to latest version) after requesting them, persist them to the internal database and then - for each of the new items coming in - Vue need to re-render large parts of the DOM. This is taking up a lot of CPU (and memory) resources in a single thread (JavaScript is single-threaded except if you use Web Workers). So all this is done in the main Browser thread, which blocks interaction to some extent.
I was able to improve the situation now by mostly (1) requesting less data in intervals (3 items, wait a bit to let other things happen, next 3 items, wait a bit etc), (2) forbidding Vue to travers through the large Catalog and Item objects (to make them reactive), which by default happens and (3) only showing a limited number of items by default and rendering them only once scrolled down ("infite scrolling").

Anyway, I think we've made good progress here and I hope we don't even need to implement Web Workers for now. Let's see how it goes. Thanks for reporting the issue and helping out
